Mathlib/Data/List/Shortlex.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2024 Hannah Fechtner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hannah Fechtner
+-/
+
+import Mathlib.Data.List.Lex
+import Mathlib.Tactic.Linarith
+import Mathlib.Order.RelClasses
+
+/-!
+# Shortlex ordering of lists.
+
+Given a relation `r` on `α`, the shortlex order on `List α` is defined by `L < M` iff
+* `L.length < M.length`
+* `L.length = M.length` and `L < M` under the lexicographic ordering over `r` on lists
+
+## Main results
+
+We show that if `r` is well-founded, so too is the shortlex order over `r`
+
+## See also
+
+Related files are:
+* `Mathlib/Data/List/Lex`: Lexicographic order on `List α`.
+* `Mathlib/Data/DFinsupp/WellFounded`: Well-foundedness of lexicographic orders on `DFinsupp` and
+  `Pi`.
+-/
+
+/-! ### shortlex ordering -/
+
+namespace List
+
+/-- Given a relation `r` on `α`, the shortlex order on `List α`, for which
+`[a0, ..., an] < [b0, ..., b_k]` if `n < k` or `n = k` and `[a0, ..., an] < [b0, ..., bk]`
+under the lexicographic order induced by `r`. -/
+def Shortlex {α : Type*} (r : α → α → Prop) : List α → List α → Prop :=
+  InvImage (Prod.Lex (· < ·) (List.Lex r)) fun a ↦ (a.length, a)
+
+namespace Shortlex
+
+variable {α : Type*} {r : α → α → Prop}
+
+/-- If a list `s` is shorter than a list `t`, then `s` is smaller than `t` under any shortlex
+order. -/
+theorem of_length_lt {s t : List α} (h : s.length < t.length) : Shortlex r s t :=
+  Prod.Lex.left _ _ h
+
+/-- If two lists `s` and `t` have the same length, `s` is smaller than `t` under the shortlex order
+over a relation `r`  when `s` is smaller than `t` under the lexicographic order over `r` -/
+theorem of_lex {s t : List α} (len_eq : s.length = t.length) (h_lex : List.Lex r s t) :
+    Shortlex r s t := by
+  apply Prod.lex_def.mpr
+  right
+  exact ⟨len_eq, h_lex⟩
+
+theorem _root_.List.shortlex_def {s t : List α} : Shortlex r s t ↔
+    s.length < t.length ∨ s.length = t.length ∧ List.Lex r s t := Prod.lex_def
+
+/-- If two lists `s` and `t` have the same length, `s` is smaller than `t` under the shortlex order
+over a relation `r` exactly when `s` is smaller than `t` under the lexicographic order over `r`.-/
+theorem _root_.List.shortlex_iff_lex {s t : List α} (h : s.length = t.length) :
+    Shortlex r s t ↔ List.Lex r s t := by
+  simp [shortlex_def, h]
+
+theorem cons_iff [IsIrrefl α r] {a : α} {s t : List α} : Shortlex r (a :: s) (a :: t) ↔
```

Comment:
```suggestion
theorem _root_.List.shortlex_cons_iff [IsIrrefl α r] {a : α} {s t : List α} : Shortlex r (a :: s) (a :: t) ↔
```
(I doubt it's useful to do `namespace Shortlex` here)

Reply:
I ignored this because the current version is I believe consistent with `List.Lex`. If you feel strongly about this, can you make a PR to clean up the Lex namespace first?

Reply:
#22935